Rename AbstractIOBuffer, it's not an abstract type

tkelman opened the pull request:
rename `AbstractIOBuffer` to `IOBufferBase`
ref discussion at https://github.com/JuliaLang/julia/pull/11554#commitcomment-16067265

~~Luckily very little code in base and no code in registered packages is using the constructor
for this type, only in dispatch (or in `precompile` calls) so this change should in theory not
be very noticeable. After we branch for release-0.5 we should consider removing the
parameterization and just call the type `IOBuffer`, since it doesn't look like anyone ever
uses any other type for the `data` field.~~

tkelman committed:
Rename AbstractIOBuffer to IOBufferBase

ref discussion at https://github.com/JuliaLang/julia/pull/11554#commitcomment-16067265
it might be better to rename this to IOBuffer but I'm not sure how to make that work

vtjnash commented:
-1 The code isn't parameterized to handle this correctly. This seems to be code churn with no benefit.

tkelman commented:
The benefit is the current `AbstractIOBuffer` isn't an abstract type at all. That's just silly naming.

TotalVerb commented:
Can't this be called `IOBuffer`? The constructor can be made to default to `Vector{UInt8}`. The only thing that should break are functions specialized on `IOBuffer` that suddenly accept all kinds of `IOBuffer{T}`s, which isn't necessarily a bad thing.

tkelman commented:
I suspect it could but I wasn't sure exactly what that would look like.

tkelman commented:
Suggestions? This seemed the least disruptive immediate resolution of the misleading naming. @stevengj you had some thoughts in february?

vtjnash commented:
If you just don't want the name, the most correct change IMO would be just `s/AbstractIOBuffer/IOBufferBase/g`

Definitely don't add an actual `abstract` type. That's completely wrong (the abstract type interface this implements is `IO`. The type itself is abstracting over the storage backing, not the interface.).

FWIW, it's probably wrong to call `AbstractIOBuffer` a concrete type. In the part of our manual that introduce parametric types, it is called an abstract type: ```What does Point by itself mean? It is an abstract type...``` (http://docs.julialang.org/en/release-0.4/manual/types/#parametric-composite-types)

tkelman commented:
We should probably be more distinct in our wording then, since `abstract` is a keyword we should avoid using that term elsewhere. I'd say something like "parameterized type family". I'll modify this to be a `@deprecate_binding` rename to `IOBufferBase` then.

vtjnash commented:
Even more confusingly, from talking with carnaval, `abstract IO` is probably actually describing a concrete type from the type-system standpoint, since it has no unbound type-parameters. Thus, if we allowed construction of non-leaf types, it might be a candidate for being an instantiatable type (cross-ref #17086 discussion as well)

stevengj commented:
Because the abstract parametric type shares the same name with the concrete parametric type for a given `T`, the usual Julia convention is to *not* use the name `Abstract`.  (We don't say `AbstractComplex` even though, by that definition, `Complex` is an abstract type.)

I tend to favor just calling this `IOBuffer`, as @TotalVerb suggests, and seeing where the cards fall.

tkelman commented:
`Complex.abstract` is false though so I don't think that's the best terminology to use for unbound parameters.

tkelman committed:
adjust indentation in doc/manual/types.rst

tkelman committed:
Remove use of 'abstract type' when discussing parametric composite types

this is a bit confusing and we don't have a great word for it,
but since Point wasn't declared with abstract here we probably
shouldn't use that term

tkelman commented:
I also like the sound of just calling the parameterized version `IOBuffer` but I'm not positive what the patch to do that correctly would look like.

ggggggggg commented:
I just looked at `iobuffer.jl` to figure out how to implement `IO` and was very confused by the name `AbstractIOBuffer` for a few minutes. I support this change, or even changing it to `IOBuffer`, though I understand from the thread that that is harder. I think there is benefit in changing the name to make the code easier to read.

StefanKarpinski commented:
The fact that AbstractIOBuffer is not an abstract type is totally ridiculous and must be fixed by 1.0.

StefanKarpinski commented:
Resolved: renaming it to `GenericIOBuffer`.
